### PR TITLE
expand HTMLComposition API

### PR DESCRIPTION
**Pull request by Mackiovello**

### CODE REVIEW

File: src/Starcounter.Extensions/HTMLComposition.cs [hunk 1]
```diff
@@ -47,19 +60,21 @@ public class HTMLComposition {
                 return 204;
             }, new HandlerOptions() { SkipRequestFilters = true });
 
-            Handle.GET("/sc/partial/composition?key={?}&ver={?}", (string key, string version) => {
+            // Returns the wanted composition by inserting a key and/or version
+            Handle.GET("/sc/partial/composition?key={?}&ver={?}&returntype={?)", (string key, string version, string returnType) => {
```
Comment (tomalec) [inline src/Starcounter.Extensions/HTMLComposition.cs]: I would change `returntype` parameter check to `Accept` request header check

File: src/Starcounter.Extensions/HTMLComposition.cs [hunk 2]
```diff
@@ -75,25 +90,80 @@ public class HTMLComposition {
...
+                Response response = new Response();
+                response.ContentType = returnType == "html" ? "text/html" : "application/json";
+                response.Body = setup.Value;
+                response.StatusCode = 200;
+                return response;
+            }, new HandlerOptions() { SkipRequestFilters = true });
 
-                    sb.Append("    ('").Append(item.Key).Append("', '").Append(item.Value).Append("')");
-                    index++;
+            // returns compositions for given partials in any set of apps
```
Comment (tomalec) [inline src/Starcounter.Extensions/HTMLComposition.cs]: in fact it returns only compositions for any set of just two apps

File: src/Starcounter.Extensions/HTMLComposition.cs [hunk 3]
```diff
@@ -75,25 +90,80 @@ public class HTMLComposition {
...
+            Handle.GET("/sc/partial/composition?partial={?}&partial={?}&returntype={?}", (string partialOne, string partialTwo, string returnType) =>
```
Comment (tomalec) [inline src/Starcounter.Extensions/HTMLComposition.cs]: I would not provide a n API just for two apps, it should be either 1 or n.

BTW, I don't think this handler will give us any benefits.

File: src/Starcounter.Extensions/HTMLComposition.cs [hunk 4]
```diff
@@ -28,10 +28,23 @@ public class HTMLComposition {
...
 
+        public static HTMLComposition GetAll() {
+            return Db.SQL<HTMLComposition>("SELECT c FROM Starcounter.HTMLComposition c");
+        }
+
+        public static HTMLComposition GetUsingAppOrPartial(string app) {
+            return Db.SQL<HTMLComposition>("SELECT c FROM Starcounter.HTMLComposition c WHERE c.Key LIKE ?", app).First;
+        }
+
+        public static HTMLComposition GetStandalonePartial(string partial) {
```
Comment (tomalec) [inline src/Starcounter.Extensions/HTMLComposition.cs]: This returns not only standalone partials, it returns any specific partial

File: src/Starcounter.Extensions/HTMLComposition.cs [hunk 5]
```diff
@@ -75,25 +90,80 @@ public class HTMLComposition {
...
-                return sb.ToString();
+                Response response = new Response();
+                response.ContentType = returnType == "html" ? "text/html" : "application/json";
+                response.Body = setup.Value;
+                response.StatusCode = 200;
+                return response;
+
             }, new HandlerOptions() { SkipRequestFilters = true });
+
+            Handle.GET("/sc/partial/composition?standalonepartial={?}&returntype={?}", (string partial, string returnType) =>
```
Comment (tomalec) [inline src/Starcounter.Extensions/HTMLComposition.cs]: As mentioned above, the way it's implemented, that handler works for any specific partial, not only standalone ones.

**Comment by tomalec** [comment]
This PR is for https://github.com/Starcounter/Starcounter/issues/3920

File: src/Starcounter.Extensions/HTMLComposition.cs (hunk 1 above)
Comment (Mackiovello) [inline src/Starcounter.Extensions/HTMLComposition.cs]: The explanation on the `Headers` property is kind of ambiguous.

Would I just put:
`Accept: "text/html"`
`Accept: "application/json"`

or something like:
`response.Header["Accept"] = "text/html"`
`response.Header["Accept"] = "application/json"`

Comment (Mackiovello) [inline src/Starcounter.Extensions/HTMLComposition.cs]: After setting the Accept header, would I still need a `ContentType`?

Comment (Mackiovello) [inline src/Starcounter.Extensions/HTMLComposition.cs]: @tomalec

**Comment by warpech** [comment]
What is the status of this?

**Comment by Mackiovello** [comment]
@warpech I think I'm done. The remaining step is to build it locally to test, if that's needed. No matter what, I would need someone to look through the code.

**Comment by warpech** [comment]
@tomalec can you pls check if this satisfies your review?

**Comment by warpech** [comment]
Ping @tomalec

File: src/Starcounter.Extensions/HTMLComposition.cs [hunk 6]
```diff
@@ -1,99 +1,180 @@
...
+            } else {
+                return Db.SQL<HTMLComposition>("SELECT c FROM Starcounter.HTMLComposition c WHERE c.Key = ? AND c.Version = ?", key, version).First;
+            }
+        }
+
+        public static HTMLComposition GetAll() {
+            return Db.SQL<QueryResultRows>("SELECT c FROM Starcounter.HTMLComposition c");
+        }
+
+        public static HTMLComposition GetUsingMergedPartials(string mergedPartials) {
```
Comment (Mackiovello) [inline src/Starcounter.Extensions/HTMLComposition.cs]: Here it's called `GetUsingMergedPartials`

File: src/Starcounter.Extensions/HTMLComposition.cs [hunk 7]
```diff
@@ -1,99 +1,180 @@
...
-                        var setup = HTMLComposition.GetUsingKeyAndVersion(key, version);
-                        if (setup != null) {
-                            setup.Delete();
-                        }
-                    }
-                });
-                return 204;
-            }, new HandlerOptions() { SkipRequestFilters = true });
-
-            Handle.GET("/sc/generate/partial/composition/{?}", (string app) => {
```
Comment (Mackiovello) [inline src/Starcounter.Extensions/HTMLComposition.cs]: I'm not sure what you're commenting on here. I left a comment below in the new code where it is name `GetUsingMergedPartials`

**Comment by warpech** [comment]
@tomalec asked me to join the review of this one. I will try to do it today.

**Comment by Mackiovello** [comment]
@alemoi why did you close this one?

**Comment by alemoi** [comment]
>  alemoi closed this 9 hours ago

Very strange! I didn't close it. Or at least don't remember of doing it. Maybe accidental?

File: src/Starcounter.Rest/Starcounter.Rest.32Bit.csproj [hunk 8]
```diff
@@ -1,10 +0,0 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (warpech) [inline src/Starcounter.Rest/Starcounter.Rest.32Bit.csproj]: Why is this file removed? Is this on purpose?

Comment (warpech) [inline src/Starcounter.Rest/Starcounter.Rest.32Bit.csproj]: Also, I suppose that the file `src/UpgradeLog.htm` is added by mistake.

File: src/Starcounter.Extensions/HTMLComposition.cs [hunk 9]
```diff
@@ -1,99 +1,184 @@
...
+                    var setup = GetUsingKeyAndVersion(key, version);
+                    if (setup == null) {
+                        setup = new HTMLComposition() {
+                            Key = key,
+                            Version = version
+                        };
+                    }
+                    setup.Value = request.Body;
+                });
+                return 204;
```
Comment (warpech) [inline src/Starcounter.Extensions/HTMLComposition.cs]: Let's not do a 204 response, because that tells nothing. More reading: http://blog.ploeh.dk/2013/04/30/rest-lesson-learned-avoid-204-responses/

Better would be to respond with something like: https://github.com/StarcounterPrefabs/SignIn/blob/a750754e9c2af6502b78d717096850ccbcef363d/src/SignIn/Api/MainHandlers.cs#L137

Maybe with status 201 Created.

**Comment by Mackiovello** [comment]
@warpech I have now responded to your comments. Please merge if it's satisfactory

**Comment by warpech** [comment]
My review is that this Pull Request does not seem needed at all. Please convince me in https://github.com/Starcounter/Starcounter/issues/3920#issuecomment-265790239. @Mackiovello thanks for being so eager to do code changes, but sometimes the best code is no code at all :)

**Comment by warpech** [comment]
Closing as we discussed that this is not really what we need. Discussion is on in #3920
